src/shared/icons.ts

```diff
@@ -1,106 +1,116 @@
+import { isCloud9 } from './extensionUtilities'
+import { memoize } from './utilities/functionUtils'
+import { getLogger } from './logger/logger'
+import { UnknownError } from './toolkitError'
+
+// Animation:
+// https://code.visualstudio.com/api/references/icons-in-labels#animation
+
+// Colors:
+// https://code.visualstudio.com/api/references/contribution-points#contributes.colors
+
+type ContributedIcon = keyof typeof packageJson.contributes.icons
+type IconPath = { light: Uri; dark: Uri } | Icon
+type IconId = `vscode-${string}` | ContributedIcon
+
+/**
+ * Gets an icon associated with the specified `id`.
+ *
+ * May return either {@link ThemeIcon} or URIs to icons. Callers should just pass the result directly
+ * to any consuming APIs.
+ *
+ * Icons contributed by the extension are prefixed based off their location in the `icons` directory.
+ *
+ * For example, `aws-s3-bucket` refers to `bucket.svg` in `icons/aws/s3`. Note that while identifiers
+ * may be derived from their source, that doesn't mean all valid identifiers have locations on disk.
+ */
+export const getIcon = memoize(resolveIconId)
+
+/**
+ * Convenient way to add icons provided by {@link getIcon} into a string.
+ *
+ * Invalid icons are simply omitted. This function also removes any leading
+ * or trailing white space from the final result.
+ *
+ * #### Example:
+ * ```ts
+ * vscode.window.setStatusBarMessage(codicon`${getIcon('vscode-clippy')} Copied to clipboard`)
+ * ```
+ */
+export function codicon(parts: TemplateStringsArray, ...components: (string | IconPath)[]): string {
+    const canUse = (sub: string | IconPath) => typeof sub === 'string' || (!isCloud9() && sub instanceof Icon)
+    const resolved = components.filter(canUse).map(String)
+
+    return parts
+        .map((v, i) => `${v}${i < resolved.length ? resolved[i] : ''}`)
+        .join('')
+        .trim()
+}
 
-    globals.iconPaths.dark.exit = context.asAbsolutePath('resources/dark/exit.svg')
-    globals.iconPaths.light.exit = context.asAbsolutePath('resources/light/exit.svg')
+/**
+ * See {@link ThemeIcon}.
+ *
+ * Used to expose the icon identifier which is otherwise hidden.
+ */
+export class Icon extends ThemeIcon {
+    public constructor(public readonly id: string) {
+        super(id)
+    }
+
+    public toString() {
+        return `$(${this.id})`
+    }
+}
 
-    cdk.iconPaths.dark.cdk = context.asAbsolutePath('resources/dark/cdk/cdk.svg')
-    cdk.iconPaths.light.cdk = context.asAbsolutePath('resources/light/cdk/cdk.svg')
+function resolveIconId(
+    id: IconId,
+    shouldUseCloud9 = isCloud9(),
+    iconsPath = globals.context.asAbsolutePath(path.join('resources', 'icons'))
+): IconPath {
+    const [namespace, ...rest] = id.split('-')
+    const name = rest.join('-')
+
+    // This 'override' logic is to support legacy use-cases, though ideally we wouldn't need it at all
+    const cloud9Override = shouldUseCloud9 ? resolvePathsSync(path.join(iconsPath, 'cloud9'), id) : undefined
```

Comment:
The override is only used for 1 file for Cloud9 but I could see people wanting icon overrides that do more than just change the color. Our common-case is monochromatic icons so that's what this PR is optimized for.